Part_10_typescript/src/task7.ts

```diff
@@ -0,0 +1,22 @@
+//7. Implement an async/await function that fetches data from an API and demonstrates error handling and handling of async operations.
+
+import axios from "axios";
+
+async function fetchData(url: string): Promise<any> {
```

Comment:
When you know the structure of the response, you can get power of typescript:

Step1. Create an interface for the response.

```ts
interface IJokeRes {
  error: boolean,
  category: string,
  type: string,
  setup: string
  delivery: string
  flags: {
    nsfw: boolean,
    religious: boolean,
    racist: boolean,
    rexist: boolean,
    explicit: boolean
  }
  safe: boolean,
  id: number,
  lang: string
}
```
Step 2. Then add generic to fetchData which will make this function more reusable:
```ts
async function fetchData<T>(url: string): Promise<T> {
  try {
    const response = await axios.get(url);
    return response.data;
  } catch (error) {
    console.error("An error occurred:", error);
    throw error;
  }
}
```

Step 3. Pass the interface into fetchData function, like so:
```ts
const apiUrl = "https://v2.jokeapi.dev/joke/Any?safe-mode";
fetchData<IJokeRes>(apiUrl)
  .then((data) => {
    console.log("Data fetched successfully:", data);
  })
  .catch((error) => {
    console.error("Error occurred while fetching data:", error);
  });
```


Then you will get hints from typescript:
![image](https://github.com/girlinitworld/CoreJS/assets/118353174/1b0de0e8-ba12-4437-a222-8a52478e1b7f)